File: tree/107.c
/**
 * 在原有的层次遍历基础上, reverse结果数组就可以了
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *columnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */

void traverse(int*** s, int** columnSizes, int* returnSize, struct TreeNode* t, int depth);

int** levelOrderBottom(struct TreeNode* root, int** columnSizes, int* returnSize) {
    int*** s = (int***)malloc(sizeof(int**));
    *s = NULL;
    if(root==NULL) return *s;
    *returnSize = 0;
    traverse(s, columnSizes, returnSize, root, 0);
    int** ret = (int**)malloc((*returnSize)*sizeof(int*));
    int* retSize = (int*)malloc((*returnSize)*sizeof(int));
    int index;
    for(int i = 0; i<*returnSize;i++)
    {
        retSize[i] = (*columnSizes)[*returnSize-1-i];
        ret[i] = (int*)malloc(sizeof(int)*((*columnSizes)[*returnSize-1-i]+1));
        ret[i] = (*s)[*returnSize-1-i];
    }
    *columnSizes = retSize;
    //columnSizes = &retSize 是不对的,只改变了副本变量的地址 没有改变源地址
    return ret;
}

void traverse(int*** s, int**columnSizes, int* returnSize, struct TreeNode* t, int depth)
{
    if(t==NULL) return;
    if(*returnSize < depth + 1)
    {
        *returnSize = depth + 1;
        *s = (int**)realloc(*s, (depth+1)*sizeof(int*));
        (*s)[depth] = NULL;
        *columnSizes = (int*)realloc(*columnSizes, (depth+1)*sizeof(int*));
        (*columnSizes)[depth] = 0;
    }
    (*s)[depth] = (int*)realloc((*s)[depth], ((*columnSizes)[depth]+1)*sizeof(int));
    (*s)[depth][(*columnSizes)[depth]] = t->val;
    ++(*columnSizes)[depth];
    traverse(s, columnSizes, returnSize, t->left, depth+1);
    traverse(s, columnSizes, returnSize, t->right, depth+1);
}

File: tree/107.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Item struct {
    level int
    t *TreeNode
}

func levelOrderBottom(root *TreeNode) [][]int {
    ret := make([][]int, 0)
    if root == nil {
        return ret
    }
    queue := make([]*Item, 0)
    queue = append(queue, &Item{0, root})
    for len(queue) > 0 {
        item := queue[0]
        queue = queue[1:]
        if item.level == len(ret) {
            ret = append(ret, []int{})
        }
        ret[item.level] = append(ret[item.level], item.t.Val)
        if item.t.Left != nil {
            queue = append(queue, &Item{item.level+1, item.t.Left})
        }
        if item.t.Right != nil {
            queue = append(queue, &Item{item.level+1, item.t.Right})
        }        
    }
    for i, j := 0, len(ret)-1; i < j; i, j = i+1, j-1 {
        ret[i], ret[j] = ret[j], ret[i]
    }
    return ret
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Item struct {
    Node *TreeNode
    Level int
}
func levelOrderBottom(root *TreeNode) [][]int {
    ret := make([][]int, 0)
    if root == nil {
        return ret
    }
    stack := make([]Item, 0)
    stack = append(stack, Item{root, 0})
    for len(stack) > 0 {
        cur := stack[0]
        stack = stack[1:]
        if len(ret) <= cur.Level {
            ret = append(ret, make([]int, 0))
        }
        ret[cur.Level] = append(ret[cur.Level], cur.Node.Val)
        if cur.Node.Left != nil {
            stack = append(stack, Item{cur.Node.Left, cur.Level+1})
        }
        if cur.Node.Right != nil {
            stack = append(stack, Item{cur.Node.Right, cur.Level+1})
        }
    }
    for i, j := 0, len(ret)-1; i < j; i, j = i+1, j-1 {
        ret[i], ret[j] = ret[j], ret[i]
    }
    return ret
}
